src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceFileService.cs

```diff
@@ -63,16 +64,17 @@ private static string CreateMutexName(string directoryName)
             return "MetadataAsSource-" + directoryName;
         }
 
-        private string GetRootPathWithGuid_NoLock()
+        private string GetRootPathWithGuid_NoLock(bool allowDecompilation)
         {
-            if (_rootTemporaryPathWithGuid == null)
+            ref string rootTemporaryPathWithGuid = ref (allowDecompilation ? ref _rootDecompiledTemporaryPathWithGuid : ref _rootMetadataTemporaryPathWithGuid);
+            if (rootTemporaryPathWithGuid == null)
             {
                 var guidString = Guid.NewGuid().ToString("N");
-                _rootTemporaryPathWithGuid = Path.Combine(_rootTemporaryPath, guidString);
+                rootTemporaryPathWithGuid = Path.Combine(_rootTemporaryPath, guidString);
                 _mutex = new Mutex(initiallyOwned: true, name: CreateMutexName(guidString));
```

Comment:
This now may be overwriting a previous Mutex created by our process; I'm guessing if the Mutex GCs then another process might clean up things under us? I believe the right fix here is to preserve a single folder, and just maybe add another layer of folders inside of it, i.e. MetadataAsSource-{guid}\Decompiled vs. MetadataAsSource-{guid}\Metadata. Or, hold onto two mutexes, but I think going back to a single directory is easier, and also scales if/when we create a third type of something to go in here...

Reply:
Two mutexes is easy with the current pattern. Yay ref locals!